pkg/operation/operation.go

```diff
@@ -280,65 +286,45 @@ func (o *Operation) InitializeSeedClients() error {
 // the already existing Operation object.
 func (o *Operation) InitializeShootClients() error {
-	if o.K8sShootClient != nil && o.ChartApplierShoot != nil {
+	if o.K8sShootClient != nil {
 		return nil
 	}
 
-	if o.Shoot.HibernationEnabled {
-		controlPlaneHibernated, err := o.controlPlaneHibernated()
-		if err != nil {
-			return err
-		}
-		// Do not initialize Shoot clients for already hibernated shoots.
-		if controlPlaneHibernated {
-			return nil
-		}
-	}
-
-	secretName := v1beta1constants.SecretNameGardener
-	// If the gardenlet runs in the same cluster like the API server of the shoot then use the internal kubeconfig
-	// and communicate internally. Otherwise, fall back to the "external" kubeconfig and communicate via the
-	// load balancer of the shoot API server.
-	addr, err := net.LookupHost(o.Shoot.ComputeInClusterAPIServerAddress(false))
+	apiServerRunning, err := o.IsAPIServerRunning()
 	if err != nil {
-		o.Logger.Warnf("service DNS name lookup of kube-apiserver failed (%+v), falling back to external kubeconfig", err)
-	} else if len(addr) > 0 {
-		secretName = v1beta1constants.SecretNameGardenerInternal
+		return err
 	}
-
-	k8sShootClient, err := kubernetes.NewClientFromSecret(o.K8sSeedClient, o.Shoot.SeedNamespace, secretName,
-		kubernetes.WithClientConnectionOptions(o.Config.ShootClientConnection.ClientConnectionConfiguration),
-		kubernetes.WithClientOptions(client.Options{
-			Scheme: kubernetes.ShootScheme,
-		}),
-	)
-	// TODO: This if-condition can be removed in a future version when all shoots were reconciled with Gardener v1.1 version.
-	if secretName == v1beta1constants.SecretNameGardenerInternal && err != nil && apierrors.IsNotFound(err) {
-		k8sShootClient, err = kubernetes.NewClientFromSecret(o.K8sSeedClient, o.Shoot.SeedNamespace, v1beta1constants.SecretNameGardener,
-			kubernetes.WithClientConnectionOptions(o.Config.ShootClientConnection.ClientConnectionConfiguration),
-			kubernetes.WithClientOptions(client.Options{
-				Scheme: kubernetes.ShootScheme,
-			}),
-		)
+	// Don't initialize clients for Shoots, that have are currently hibernated or in the process of being created or deleted.
+	if !apiServerRunning {
+		return nil
 	}
+
+	shootClient, err := o.ClientMap.GetClient(context.TODO(), keys.ForShoot(o.Shoot.Info))
 	if err != nil {
 		return err
 	}
-	o.K8sShootClient = k8sShootClient
+	o.K8sShootClient = shootClient
 
-	o.ChartApplierShoot = k8sShootClient.ChartApplier()
 	return nil
 }
 
-func (o *Operation) controlPlaneHibernated() (bool, error) {
-	replicaCount, err := common.CurrentReplicaCount(o.K8sSeedClient.Client(), o.Shoot.SeedNamespace, v1beta1constants.DeploymentNameKubeAPIServer)
-	if err != nil {
+func (o *Operation) IsAPIServerRunning() (bool, error) {
+	deployment := &appsv1.Deployment{}
+	if err := o.K8sSeedClient.DirectClient().Get(context.TODO(), kutil.Key(o.Shoot.SeedNamespace, v1beta1constants.DeploymentNameKubeAPIServer), deployment); err != nil {
```

Comment:
Why you use the direct client here?

Reply:
Basically just to make sure, we're not reading from a stale cache, when checking if we should initialize a shoot client in the care controller. But no strong opinion here.

Reply:
Hm ok.. agreed, no other opinion either. Maybe you can add a small comment that explains why the direct client is used. 👍